queue/abstract.lua

```diff
@@ -375,49 +389,45 @@ local function make_self(driver, space, tube_name, tube_type, tube_id, opts)
 end
 
 --- Release all session tasks.
-local function release_session_tasks(connection_id)
-    while true do
-        local task = box.space._queue_taken.index.pk:min{connection_id}
-        if task == nil or task[1] ~= connection_id then
-            break
-        end
+local function release_session_tasks(session_uuid)
+    local taken_tasks = box.space._queue_taken_2.index.uuid:select{session_uuid}
 
-        local tube = box.space._queue.index.tube_id:get{task[2]}
+    for _, task in pairs(taken_tasks) do
+        local tube = box.space._queue.index.tube_id:get{task[1]}
         if tube == nil then
-            log.error("Inconsistent queue state: tube %d not found", task[2])
-            box.space._queue_taken.index.task:delete{task[2], task[3]}
+            log.error("Inconsistent queue state: tube %d not found", task[1])
+            box.space._queue_taken_2.index.task:delete{task[1], task[2]}
         else
-            log.warn("Consumer %s disconnected, release task %s(%s)",
-                connection_id, task[3], tube[1])
-
-            tube_release_internal(queue.tube[tube[1]], task[3], nil,
-                                  connection_id)
+            log.warn("Session %s closed, release task %s(%s)",
+                uuid.frombin(session_uuid):str(), task[2], tube[1])
+            tube_release_internal(queue.tube[tube[1]], task[2], nil,
+                session_uuid)
         end
     end
 end
 
 function method._on_consumer_disconnect()
-    local waiter, fb, task, tube, id
-    id = connection.id()
+    local conn_id = connection.id()
+
     -- wakeup all waiters
     while true do
-        waiter = box.space._queue_consumers.index.pk:min{id}
+        local waiter = box.space._queue_consumers.index.pk:min{conn_id}
```

Comment:
Just to understand the idea: consumers are still identified by a connection id rather than queue session id? Can you explain, why it is different from identifying a task ownership?

Reply:
Several consumers can work in the same session with the same tasks. For example, one consumer can `take` a task and another consumer (in the same session) can `ack` this task.